### PR TITLE
Multithread cp for directories

### PR DESCRIPTION
**Pull request by IanButterworth**
I don't know whether multithreading like this in Base is allowed, but this appears to speed up `cp` on directories on my setup significantly (6-thread SSD macbook, MacOS)

Master
```julia
julia> @time cp("/Users/ian/.julia/registries/General", mktempdir(), force=true);
  6.647516 seconds (287.42 k allocations: 33.990 MiB)

julia> @time cp("/Users/ian/.julia/registries/General", mktempdir(), force=true);
  7.410636 seconds (287.42 k allocations: 33.990 MiB, 0.43% gc time)

julia> @time cp("/Users/ian/.julia/registries/General", mktempdir(), force=true);
  6.996436 seconds (287.43 k allocations: 33.991 MiB, 1.87% gc time)
```

This PR
```julia
julia> @time cp("/Users/ian/.julia/registries/General", mktempdir(), force=true);
  1.514411 seconds (409.63 k allocations: 59.738 MiB, 0.36% gc time)

julia> @time cp("/Users/ian/.julia/registries/General", mktempdir(), force=true);
  1.621347 seconds (409.63 k allocations: 59.738 MiB, 0.35% gc time)

julia> @time cp("/Users/ian/.julia/registries/General", mktempdir(), force=true);
  1.682619 seconds (409.61 k allocations: 59.738 MiB, 0.55% gc time)


# Parity check

julia> d = mktempdir()
"/var/folders/_6/1yf6sj0950vcg4t91m9ltb5w0000gn/T/jl_bY80nJ"

julia> @time cp("/Users/ian/.julia/registries/General", d, force=true);
  1.688429 seconds (409.65 k allocations: 59.739 MiB, 0.19% gc time)

julia> Pkg.GitTools.tree_hash("/Users/ian/.julia/registries/General") == Pkg.GitTools.tree_hash(d)
true
```

### CODE REVIEW

**Comment by IanButterworth** [comment]
The typical load-balanced `ntasks=Threads.nthreads()+1` default is backed up by a pretty repeatable minima at `ntasks=7` on a 6 thread system

```julia
julia> for ntasks in 1:20
           t = @elapsed cp("/Users/ian/.julia/registries/General", mktempdir(), force=true,  ntasks=ntasks)
           @show ntasks, t
       end

(ntasks, t) = (1, 5.648116088)
(ntasks, t) = (2, 2.312361291)
(ntasks, t) = (3, 2.087711314)
(ntasks, t) = (4, 1.805751717)
(ntasks, t) = (5, 1.880124248)
(ntasks, t) = (6, 1.830967428)
(ntasks, t) = (7, 1.798523184)
(ntasks, t) = (8, 1.889625656)
(ntasks, t) = (9, 2.081643231)
(ntasks, t) = (10, 2.062828056)
(ntasks, t) = (11, 1.95200327)
(ntasks, t) = (12, 1.954158919)
(ntasks, t) = (13, 1.987198201)
(ntasks, t) = (14, 2.155334364)
(ntasks, t) = (15, 2.048653877)
(ntasks, t) = (16, 2.170309276)
(ntasks, t) = (17, 2.155259291)
(ntasks, t) = (18, 2.246891216)
(ntasks, t) = (19, 2.180693019)
(ntasks, t) = (20, 2.14442822)
```

**Comment by IanButterworth** [comment]
In a worst-case, if users already spawn a `cp` action per-thread, the behavior is still performant.

i.e. here there are 36 concurrent `sendfile` tasks, and the average time per `cp` is ~1.6s (9.4s/6)
```julia
julia> Threads.@threads for _ in 1:Threads.nthreads()
           @time cp("/Users/ian/.julia/registries/General", mktempdir(), force=true)
       end
  9.087354 seconds (2.28 M allocations: 332.051 MiB, 0.45% gc time)
  9.407884 seconds (2.37 M allocations: 345.358 MiB, 0.43% gc time)
  9.562321 seconds (2.41 M allocations: 351.945 MiB, 0.42% gc time)
  9.618445 seconds (2.43 M allocations: 354.279 MiB, 0.42% gc time)
  9.629988 seconds (2.43 M allocations: 354.760 MiB, 0.42% gc time)
  9.774446 seconds (2.46 M allocations: 358.397 MiB, 0.88% gc time)
```

File: base/file.jl [hunk 1]
```diff
@@ -329,33 +329,55 @@ function cptree(src::String, dst::String; force::Bool=false,
...
     elseif isdir(src)
-        cptree(src, dst; force=force, follow_symlinks=follow_symlinks)
+        if ntasks > 1
+            task_sema = Base.Semaphore(ntasks)
+            cptree(src, dst; force=force, follow_symlinks=follow_symlinks, task_sema=task_sema)
+            while task_sema.curr_cnt > 0 # wait until all copy tasks are done
+                lock(task_sema.cond_wait) do
+                    wait(task_sema.cond_wait)
+                end
+            end
```
Comment (IanButterworth) [inline base/file.jl:377]: Is this a reliable way to ensure all tasks are done?

**Comment by vtjnash** [comment]
Seems worthwhile then. At the lowest layer, the last parameter to libuv `uv_fs_*` functions is whether to run this asynchronously (mostly with threads, since the existence of an async kernel API is still quite new), which is currently turned off. So we might want to toggle it there instead?

**Comment by IanButterworth** [comment]
If we did it there, would we still need to semaphore? My thinking was that semaphoring and controlling tasks at the julia level would be necessary in case the directory had a huge number of files, so spawning a limited number of tasks at a time would protect against overflow?

**Comment by IanButterworth** [comment]
Results from an ubuntu VM with 4 threads
Master
```
julia> @time cp("/home/parallels/.julia/registries/General", mktempdir(), force=true);
  0.845588 seconds (288.64 k allocations: 33.191 MiB)

julia> @time cp("/home/parallels/.julia/registries/General", mktempdir(), force=true);
  0.931749 seconds (288.64 k allocations: 33.191 MiB, 3.83% gc time)
```

This PR
```
julia> @time cp("/home/parallels/.julia/registries/General", mktempdir(), force=true);
  0.537254 seconds (411.42 k allocations: 42.860 MiB)

julia> @time cp("/home/parallels/.julia/registries/General", mktempdir(), force=true);
  0.453760 seconds (411.49 k allocations: 42.862 MiB, 2.75% gc time)
```

**Comment by vtjnash** [comment]
Yes, it could be used without spawning Tasks, letting it manage the queue internally

**Comment by IanButterworth** [comment]
Ok. I just took a look and couldn't figure out how I'd go about implementing that

**Comment by IanButterworth** [comment]
> At the lowest layer, the last parameter to libuv uv_fs_* functions is whether to run this asynchronously

I took a look and the last param is a callback function. I couldn't figure out how to reorganize things to handle and synchronize the callbacks

**Comment by IanButterworth** [comment]
@vtjnash just checking this implementation isn't worth merging? Perhaps the other could come later?

**Comment by vtjnash** [comment]
Yes, it generally would be better to use the IO load balancer, rather than simultaneously oversubscribing the IO queue, and blocking the CPU queue.

**Comment by IanButterworth** [comment]
I took another look but still don't understand how it would work.

If the outer `sendfile` currently opens and closes the `src` and `dst`, is your suggestion that the close goes into the callback for `jl_fs_sendfile` within the inner `sendfile`?
But the inner `sendfile` seems to send the file in chunks, so the callback would have to be aware if it was on the last chunk, which could be nontrivial if the chunks ended up being finished out of order given async.

Also, couldn't that approach result in too many files open?

Outer
https://github.com/JuliaLang/julia/blob/fc69c9aef38a1b53324abe9a715514f9716d270b/base/file.jl#L963-L983

Inner
https://github.com/JuliaLang/julia/blob/fc69c9aef38a1b53324abe9a715514f9716d270b/base/filesystem.jl#L116-L129

**Comment by User-764Q** [comment]
Bump

**Review by vtjnash — CHANGES_REQUESTED** [review]
Need to switch this to using the IO workqueue in libuv and async (or a bounded Channel) to feed it.

**Comment by IanButterworth** [comment]
Closing this as stale, and it seems it's better suited to someone more familiar with libuv

**Pull request closed without merging**